tiledb/sm/filesystem/mem_filesystem.cc

```diff
@@ -0,0 +1,423 @@
+        string("ls not possible, in-memory filesystem is not initialized")));
+  }
+
+  vector<string> tokens = tokenize(path);
+  FSNode* cur = root_.get();
+  string dir;
+  for (auto& token : tokens) {
+    dir = dir + token + "/";
+    cur = cur->children_[token].get();
+  }
+
+  *paths = cur->ls(dir);
+  return Status::Ok();
+}
+
+Status MemFilesystem::move(
+    const std::string& old_path, const std::string& new_path) const {
+  assert(root_);
+  FSNode* old_parent_node = lookup_parent_node(old_path);
+  if (old_parent_node == nullptr) {
+    return LOG_STATUS(
+        Status::MemFSError(string("Move failed, file not found: " + old_path)));
+  }
+
+  vector<string> old_tokens = tokenize(old_path);
+  vector<string> new_tokens = tokenize(new_path);
+  string new_dir;
+  for (unsigned long i = 0; i < new_tokens.size() - 1; i++) {
+    new_dir = new_dir + new_tokens[i] + "/";
+  }
+
+  FSNode* new_parent_node = lookup_node(new_dir);
+  if (new_parent_node == nullptr) {
+    create_dir(new_dir);
+    new_parent_node = lookup_node(new_dir);
+  }
+
+  new_parent_node->children_[new_tokens.back()] =
+      std::move(old_parent_node->children_[old_tokens.back()]);
+  new_parent_node->children_[new_tokens.back()]->name_ = new_tokens.back();
+  old_parent_node->children_.erase(old_tokens.back());
+
+  return Status::Ok();
+}
+
+Status MemFilesystem::read(
+    const string& path,
+    const uint64_t offset,
+    void* buffer,
+    const uint64_t nbytes) const {
+  FSNode* node = lookup_node(path);
+  if ((node == nullptr) || node->is_dir()) {
+    return LOG_STATUS(Status::MemFSError(
+        string("File not found, read failed for : " + path)));
+  }
+
+  return ((File*)node)->read(offset, buffer, nbytes);
+}
+
+Status MemFilesystem::remove(const string& path, bool is_dir) const {
+  assert(root_);
+  if (path == "") {
+    return LOG_STATUS(
+        Status::MemFSError(string("Cannot remove the root directory")));
+  }
+
+  vector<string> tokens = tokenize(path);
+  auto cur = root_.get();
+  FSNode* parent = nullptr;
+  for (const auto& token : tokens) {
+    if (cur->has_child(token)) {
+      parent = cur;
+      cur = cur->children_[token].get();
+    } else {
+      return LOG_STATUS(Status::MemFSError(
+          string("File not found, remove failed for : " + path)));
+    }
+  }
+
+  if (cur->is_dir() != is_dir) {
+    return LOG_STATUS(
+        Status::MemFSError(string("Remove failed, wrong file type")));
+  }
+
+  if (parent) {
+    parent->children_.erase(tokens.back());
+  }
+  return Status::Ok();
+};
+
+Status MemFilesystem::touch(const string& path) const {
+  assert(root_);
+  vector<string> tokens = tokenize(path);
+  FSNode* cur = root_.get();
+  for (unsigned long i = 0; i < tokens.size() - 1; i++) {
+    auto token = tokens[i];
+    if (!cur->has_child(token)) {
+      cur->children_[token] = unique_ptr<Directory>(new Directory(token));
+    }
+    cur = cur->children_[token].get();
```

Comment:
check that this is a directory, otherwise we might be able to create a file as a child of another file

Reply:
right, thanks for the comment. Actually looking at it it's even more problematic as if I abort at any point because the node is a file, then I would need to go back and delete all the directories I have created so far. I think that's too much complexity, and we can live with simply aborting if the parent directory of the file we are creating does not exist (avoid to create it on the fly as I was doing so far). I experimented a bit and I think this is the classic behavior of *touch* in posix anyway. Is this acceptable for you?

Reply:
In the current implementation, it looks like intermediate directories are not created. This is OK.